File: go/types/set.go
// Copyright 2016 Attic Labs, Inc. All rights reserved.
// Licensed under the Apache License, version 2.0:
// http://www.apache.org/licenses/LICENSE-2.0

package types

import (
	"fmt"
	"sort"

	"github.com/attic-labs/noms/go/d"
)

type Set struct {
	orderedSequence
}

func newSet(seq orderedSequence) Set {
	return Set{seq}
}

func NewSet(vrw ValueReadWriter, v ...Value) Set {
	data := buildSetData(v)
	ch := newEmptySetSequenceChunker(vrw)

	for _, v := range data {
		ch.Append(v)
	}

	return newSet(ch.Done().(orderedSequence))
}

// NewStreamingSet takes an input channel of values and returns a output
// channel that will produce a finished Set. Values that are sent to the input
// channel must be in Noms sortorder, adding values to the input channel
// out of order will result in a panic. Once the input channel is closed
// by the caller, a finished Set will be sent to the output channel. See
// graph_builder.go for building collections with values that are not in order.
func NewStreamingSet(vrw ValueReadWriter, vChan <-chan Value) <-chan Set {
	return newStreamingSet(vrw, vChan, func(vrw ValueReadWriter, vChan <-chan Value, outChan chan<- Set) {
		go readSetInput(vrw, vChan, outChan)
	})
}

type streamingSetReadFunc func(vrw ValueReadWriter, vChan <-chan Value, outChan chan<- Set)

func newStreamingSet(vrw ValueReadWriter, vChan <-chan Value, readFunc streamingSetReadFunc) <-chan Set {
	d.PanicIfTrue(vrw == nil)
	outChan := make(chan Set, 1)
	readFunc(vrw, vChan, outChan)
	return outChan
}

func readSetInput(vrw ValueReadWriter, vChan <-chan Value, outChan chan<- Set) {
	defer close(outChan)
	ch := newEmptySetSequenceChunker(vrw)
	var lastV Value
	for v := range vChan {
		d.PanicIfTrue(v == nil)
		if lastV != nil {
			d.PanicIfFalse(lastV.Less(v))
		}
		lastV = v
		ch.Append(v)
	}
	outChan <- newSet(ch.Done().(orderedSequence))
}

// Diff computes the diff from |last| to |m| using the top-down algorithm,
// which completes as fast as possible while taking longer to return early
// results than left-to-right.
func (s Set) Diff(last Set, changes chan<- ValueChanged, closeChan <-chan struct{}) {
	if s.Equals(last) {
		return
	}
	orderedSequenceDiffTopDown(last.orderedSequence, s.orderedSequence, changes, closeChan)
}

// DiffHybrid computes the diff from |last| to |s| using a hybrid algorithm
// which balances returning results early vs completing quickly, if possible.
func (s Set) DiffHybrid(last Set, changes chan<- ValueChanged, closeChan <-chan struct{}) {
	if s.Equals(last) {
		return
	}
	orderedSequenceDiffBest(last.orderedSequence, s.orderedSequence, changes, closeChan)
}

// DiffLeftRight computes the diff from |last| to |s| using a left-to-right
// streaming approach, optimised for returning results early, but not
// completing quickly.
func (s Set) DiffLeftRight(last Set, changes chan<- ValueChanged, closeChan <-chan struct{}) {
	if s.Equals(last) {
		return
	}
	orderedSequenceDiffLeftRight(last.orderedSequence, s.orderedSequence, changes, closeChan)
}

func (s Set) asSequence() sequence {
	return s.orderedSequence
}

// Value interface
func (s Set) Value() Value {
	return s
}

func (s Set) WalkValues(cb ValueCallback) {
	iterAll(s, func(v Value, idx uint64) {
		cb(v)
	})
}

func (s Set) First() Value {
	cur := newCursorAt(s.orderedSequence, emptyKey, false, false)
	if !cur.valid() {
		return nil
	}
	return cur.current().(Value)
}

func (s Set) At(idx uint64) Value {
	if idx >= s.Len() {
		panic(fmt.Errorf("Out of bounds: %d >= %d", idx, s.Len()))
	}

	cur := newCursorAtIndex(s.orderedSequence, idx)
	return cur.current().(Value)
}

func (s Set) Has(v Value) bool {
	cur := newCursorAtValue(s.orderedSequence, v, false, false)
	return cur.valid() && cur.current().(Value).Equals(v)
}

type setIterCallback func(v Value) bool

func (s Set) Iter(cb setIterCallback) {
	cur := newCursorAt(s.orderedSequence, emptyKey, false, false)
	cur.iter(func(v interface{}) bool {
		return cb(v.(Value))
	})
}

type setIterAllCallback func(v Value)

func (s Set) IterAll(cb setIterAllCallback) {
	iterAll(s, func(v Value, idx uint64) {
		cb(v)
	})
}

func (s Set) Iterator() SetIterator {
	return s.IteratorAt(0)
}

func (s Set) IteratorAt(idx uint64) SetIterator {
	return &setIterator{
		cursor: newCursorAtIndex(s.orderedSequence, idx),
		s:      s,
	}
}

func (s Set) IteratorFrom(val Value) SetIterator {
	return &setIterator{
		cursor: newCursorAtValue(s.orderedSequence, val, false, false),
		s:      s,
	}
}

func (s Set) Edit() *SetEditor {
	return NewSetEditor(s)
}

func buildSetData(values ValueSlice) ValueSlice {
	if len(values) == 0 {
		return ValueSlice{}
	}

	uniqueSorted := make(ValueSlice, 0, len(values))
	sort.Stable(values)
	last := values[0]
	for i := 1; i < len(values); i++ {
		v := values[i]
		if !v.Equals(last) {
			uniqueSorted = append(uniqueSorted, last)
		}
		last = v
	}

	return append(uniqueSorted, last)
}

func makeSetLeafChunkFn(vrw ValueReadWriter) makeChunkFn {
	return func(level uint64, items []sequenceItem) (Collection, orderedKey, uint64) {
		d.PanicIfFalse(level == 0)
		setData := make([]Value, len(items), len(items))

		var lastValue Value
		for i, item := range items {
			v := item.(Value)
			d.PanicIfFalse(lastValue == nil || lastValue.Less(v))
			lastValue = v
			setData[i] = v
		}

		set := newSet(newSetLeafSequence(vrw, setData...))
		var key orderedKey
		if len(setData) > 0 {
			key = newOrderedKey(setData[len(setData)-1])
		}

		return set, key, uint64(len(items))
	}
}

func newEmptySetSequenceChunker(vrw ValueReadWriter) *sequenceChunker {
	return newEmptySequenceChunker(vrw, makeSetLeafChunkFn(vrw), newOrderedMetaSequenceChunkFn(SetKind, vrw), hashValueBytes)
}

File: js/noms/src/set.js
// Copyright 2016 Attic Labs, Inc. All rights reserved.
// Licensed under the Apache License, version 2.0:
// http://www.apache.org/licenses/LICENSE-2.0

// @flow

import type {ValueReader} from './value-store.js';
import type {makeChunkFn} from './sequence-chunker.js';
import type Value, {ValueCallback} from './value.js'; // eslint-disable-line no-unused-vars
import {AsyncIterator} from './async-iterator.js';
import {chunkSequence, chunkSequenceSync} from './sequence-chunker.js';
import Collection from './collection.js';
import {compare, equals} from './compare.js';
import {invariant} from './assert.js';
import Sequence, {OrderedKey} from './sequence.js';
import {newCursorAtIndex} from './indexed-sequence.js';
import {
  newOrderedMetaSequenceChunkFn,
} from './meta-sequence.js';
import {
  OrderedSequenceCursor,
  OrderedSequenceIterator,
  newCursorAt,
  newCursorAtValue,
} from './ordered-sequence.js';
import diff from './ordered-sequence-diff.js';
import {makeSetType, makeUnionType, getTypeOfValue} from './type.js';
import {removeDuplicateFromOrdered} from './map.js';
import {Kind} from './noms-kind.js';
import {hashValueBytes} from './rolling-value-hasher.js';

function newSetLeafChunkFn<T:Value>(vr: ?ValueReader): makeChunkFn<any, any> {
  return (items: Array<T>) => {
    const key = new OrderedKey(items.length > 0 ? items[items.length - 1] : false);
    const seq = newSetLeafSequence(vr, items);
    const ns = Set.fromSequence(seq);
    return [ns, key, items.length];
  };
}

function buildSetData<T: Value>(values: Array<any>): Array<T> {
  values = values.slice();
  values.sort(compare);
  return removeDuplicateFromOrdered(values, compare);
}

export function newSetLeafSequence<K: Value>(
    vr: ?ValueReader, items: K[]): SetLeafSequence<any> {
  const t = makeSetType(makeUnionType(items.map(getTypeOfValue)));
  return new SetLeafSequence(vr, t, items);
}

export default class Set<T: Value> extends Collection<Sequence<any>> {
  constructor(values: Array<T> = []) {
    const seq = chunkSequenceSync(
        buildSetData(values),
        newSetLeafChunkFn(null),
        newOrderedMetaSequenceChunkFn(Kind.Set, null),
        hashValueBytes);
    super(seq);
  }

  walkValues(vr: ValueReader, cb: ValueCallback): Promise<void> {
    return this.forEach(v => cb(v));
  }

  async has(key: T): Promise<boolean> {
    const cursor = await newCursorAtValue(this.sequence, key);
    return cursor.valid && equals(cursor.getCurrentKey().value(), key);
  }

  async _firstOrLast(last: boolean): Promise<?T> {
    const cursor = await newCursorAt(this.sequence, null, false, last);
    return cursor.valid ? cursor.getCurrent() : null;
  }

  first(): Promise<?T> {
    return this._firstOrLast(false);
  }

  last(): Promise<?T> {
    return this._firstOrLast(true);
  }

  // Returns the value at index `idx`, as per the (stable) ordering of noms
  // values in this set.
  at(idx: number): Promise<T> {
    invariant(idx >= 0 && idx < this.size);
    return newCursorAtIndex(this.sequence, idx).then(cur => cur.getCurrent());
  }

  async forEach(cb: (v: T) => ?Promise<any>): Promise<void> {
    const cursor = await newCursorAt(this.sequence, null, false, false, true);
    const promises = [];
    await cursor.iter(v => {
      promises.push(cb(v));
      return false;
    });
    await Promise.all(promises);
  }

  iterator(): AsyncIterator<T> {
    return new OrderedSequenceIterator(newCursorAt(this.sequence, null, false, false, true));
  }

  iteratorAt(v: T): AsyncIterator<T> {
    return new OrderedSequenceIterator(newCursorAtValue(this.sequence, v, false, false, true));
  }

  _splice(cursor: OrderedSequenceCursor<any, any>, insert: Array<T>, remove: number)
      : Promise<Set<T>> {
    const vr = this.sequence.vr;
    return chunkSequence(cursor, vr, insert, remove, newSetLeafChunkFn(vr),
                         newOrderedMetaSequenceChunkFn(Kind.Set, vr),
                         hashValueBytes).then(s => Set.fromSequence(s));
  }

  async add(value: T): Promise<Set<T>> {
    const cursor = await newCursorAtValue(this.sequence, value, true);
    if (cursor.valid && equals(cursor.getCurrentKey().value(), value)) {
      return this;
    }

    return this._splice(cursor, [value], 0);
  }

  async delete(value: T): Promise<Set<T>> {
    const cursor = await newCursorAtValue(this.sequence, value);
    if (cursor.valid && equals(cursor.getCurrentKey().value(), value)) {
      return this._splice(cursor, [], 1);
    }

    return this;
  }

  // TODO: Find some way to return a Set.
  async map<S>(cb: (v: T) => (Promise<S> | S)): Promise<Array<S>> {
    const cursor = await newCursorAt(this.sequence, null);
    const values = [];
    await cursor.iter(v => {
      values.push(cb(v));
      return false;
    });

    return Promise.all(values);
  }

  get size(): number {
    return this.sequence.numLeaves;
  }

  /**
   * Returns a 2-tuple [added, removed] sorted values.
   */
  diff(from: Set<T>): Promise<[Array<T> /* added */, Array<T> /* removed */]> {
    return diff(from.sequence, this.sequence).then(([added, removed, modified]) => {
      invariant(modified.length === 0);
      return [added, removed];
    });
  }
}

export class SetLeafSequence<K: Value> extends Sequence<K> {}
